### PR TITLE
fmapRouter.

### PR DESCRIPTION
**Pull request by fisx**
this function is a waste product of something we ended up doing differently, but @jkarni and I but like it a lot.  If you want to have it, I can write the test code.  If you think we should wait for use case, feel free to close this pr.

### CODE REVIEW

**Comment by alpmestan** [comment]
Yeah, could be useful. Would you mind adding some haddocks for it? And fixing [this error](https://travis-ci.org/haskell-servant/servant/jobs/84935157#L235) spotted by travis. Just an `#ifdef`'d `import Applicative` to add =)

**Comment by fisx** [comment]
I'm making a mess again...  (-:

I'll squash this tomorrow if travis and you guys have no more complaints.

File: servant-server/src/Servant/Server/Internal/Router.hs [hunk 1]
```diff
@@ -1,25 +1,40 @@
...
+  | LeafRouter    a
       -- ^ to be used for routes that match an empty path
   | Choice        Router Router
       -- ^ left-biased choice between two routers
+  deriving Functor
+
+-- | Apply a function to the result of a router in functor style.  The result contains the failure
+-- cases so one use case is to turn failures into middleware response values with appropriate status
+-- codes, message bodies, etc.
+tweakResponse :: (RouteResult Response -> RouteResult Response) -> Router -> Router
```
Comment (jkarni) [inline servant-server/src/Servant/Server/Internal/Router.hs]: We should make sure to export this from a non-internal module.

File: servant-server/src/Servant/Server/Internal/Router.hs [hunk 2]
```diff
@@ -1,25 +1,40 @@
...
   | DynamicRouter (Text -> Router)
       -- ^ first path component used for lookup and removed afterwards
-  | LeafRouter    RoutingApplication
+  | LeafRouter    a
       -- ^ to be used for routes that match an empty path
   | Choice        Router Router
       -- ^ left-biased choice between two routers
+  deriving Functor
+
+-- | Apply a function to the result of a router in functor style.  The result contains the failure
```
Comment (jkarni) [inline servant-server/src/Servant/Server/Internal/Router.hs]: This haddock isn't very clear to me. "Functor style" doesn't seem helpful, and it's not clear to me why *middleware response values* means, or why only failures are mentioned. Why not just "Apply the transformation to the response of a `Router`"?

**Comment by alpmestan** [comment]
LGTM.

**Comment by kosmikus** [comment]
Even if it's too late, LGTM as well :)

**Pull request merged**